lib/Containers/MerkleTree.cpp

```diff
@@ -262,199 +175,224 @@ MerkleTree<Hasher, BranchingBits>::fromBottomMostCompressed(std::string_view buf
-  for (std::size_t d = maxDepth - 1; /*d >= 0*/; --d) {
-    std::uint64_t prevOffset = (d == 0 ? 0 : nodeCountUpToDepth(d - 1));
-    std::uint64_t ourOffset = nodeCountUpToDepth(d);
-    std::uint64_t const ourEnd = nodeCountUpToDepth(d + 1);
-
-    while (ourOffset < ourEnd) {
-      std::uint64_t count = 0;
-      std::uint64_t hash = 0;
-      for (std::size_t index = 0; index < (1 << BranchingBits); ++index) {
-        Node const& src = tree->node(ourOffset++);
-        count += src.count;
-        hash ^= src.hash;
-      }
-      Node& dst = tree->node(prevOffset);
-      dst.count = count;
-      dst.hash = hash;
 
-      ++prevOffset;
-    }
-    if (d == 0) { 
-      break;
-    }
+  if (summaryCount != totalCount || summaryHash != totalHash) {
+    throw std::invalid_argument("invalid compressed tree summary data");
   }
 
-#ifdef PARANOID_TREE_CHECKS
-  tree->checkInternalConsistency();
-#endif
-
+  // write summary node
+  Node& summary = tree->meta().summary;
+  summary = { totalCount, totalHash };
+  
   return tree;
 }
 
 template <typename Hasher, std::uint64_t const BranchingBits>
 std::unique_ptr<MerkleTree<Hasher, BranchingBits>>
 MerkleTree<Hasher, BranchingBits>::deserialize(velocypack::Slice slice) {
-  std::unique_ptr<MerkleTree<Hasher, BranchingBits>> tree;
-
   if (!slice.isObject()) {
-    return tree;
+    return nullptr;
   }
 
   velocypack::Slice read = slice.get(StaticStrings::RevisionTreeVersion);
   if (!read.isNumber() || read.getNumber<std::uint8_t>() != ::CurrentVersion) {
-    return tree;
+    return nullptr;
   }
 
   read = slice.get(StaticStrings::RevisionTreeBranchingFactor);
   if (!read.isNumber() || read.getNumber<std::uint64_t>() != BranchingFactor) {
-    return tree;
+    return nullptr;
   }
 
   read = slice.get(StaticStrings::RevisionTreeMaxDepth);
   if (!read.isNumber()) {
-    return tree;
+    return nullptr;
   }
-  std::uint64_t maxDepth = read.getNumber<std::uint64_t>();
+  std::uint64_t depth = read.getNumber<std::uint64_t>();
 
   read = slice.get(StaticStrings::RevisionTreeRangeMax);
   if (!read.isString()) {
-    return tree;
+    return nullptr;
   }
   velocypack::ValueLength l;
   char const* p = read.getString(l);
   std::uint64_t rangeMax = basics::HybridLogicalClock::decodeTimeStamp(p, l);
   if (rangeMax == std::numeric_limits<std::uint64_t>::max()) {
-    return tree;
+    return nullptr;
   }
 
   read = slice.get(StaticStrings::RevisionTreeRangeMin);
   if (!read.isString()) {
-    return tree;
+    return nullptr;
   }
   p = read.getString(l);
   std::uint64_t rangeMin = basics::HybridLogicalClock::decodeTimeStamp(p, l);
   if (rangeMin == std::numeric_limits<std::uint64_t>::max()) {
-    return tree;
+    return nullptr;
+  }
+
+  read = slice.get(StaticStrings::RevisionTreeInitialRangeMin);
+  if (!read.isString()) {
+    return nullptr;
+  }
+  p = read.getString(l);
+  std::uint64_t initialRangeMin = basics::HybridLogicalClock::decodeTimeStamp(p, l);
```

Comment:
This code could be simplified since there is also a `decodeTimeStamp` function that takes a slice.

Reply:
Applied as suggested. Thanks!